source/sliding_puzzle.py

```diff
@@ -181,6 +182,38 @@ def shuffle(self) -> None:
         # Clear a random piece
         random.choice(random.choice(self.rows)).clear()
 
+    def shuffle2(self, difficulty: int = 50) -> None:
+        """Shuffle the puzzle by randomizing the order of the pieces
+
+        :param difficulty: approx. maximum number of iterations
+        """
+        random.choice(random.choice(self.rows)).clear()
+        clear_id = self._get_empty_piece_position()
+        clear_id = [clear_id.x, clear_id.y]
+        hash_list = [[v * self.dim[0] + i for i in range(self.dim[0])] for v in range(self.dim[1])]
+        iters = random.randint(difficulty//2, difficulty)
+
+        for i in range(iters):
```

Comment:
So this loop generates instructions for that second loop to do?

Reply:
this loop executives the movements
while the next loop updates the rows